### PR TITLE
EASY-1167: zet springfield (SF_*) variabelen klaar

### PR DESCRIPTION
**Pull request by rvanheest**
fixes EASY-1167

#### When applied it will
* change the examples to satisfy the new A/V requirements
* test all requirements in the preconditions
* add the `SF_*` fields to the `deposit.properties`

@DANS-KNAW/easy for review

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/easy/multideposit/Headers.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+package nl.knaw.dans.easy.multideposit
+
+import nl.knaw.dans.easy.multideposit.DDM._
+
+object Headers {
+  val springfieldHeaders = List("SF_DOMAIN", "SF_USER", "SF_COLLECTION", "SF_ACCESSIBILITY")
+  val audioVideoHeaders = List("AV_FILE", "AV_FILE_TITLE", "AV_SUBTITLES", "AV_SUBTITLES_LANGUAGE")
+  val administrativeHeaders = List("DEPOSITOR_ID")
+
+  val validHeaders: List[DatasetID] = {
```
Comment (PaulBoon) [inline src/main/scala/nl/knaw/dans/easy/multideposit/Headers.scala]: DatasetID is a String, but the name makes no sense for a list of header column names in my opinion.

Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/Headers.scala]: great catch!

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala [hunk 2]
```diff
@@ -35,66 +34,94 @@ case class AddPropertiesToDeposit(row: Int, entry: (DatasetID, Dataset))(implici
...
+    val sf = getSpringfieldData
+    val props: Map[String, Option[String]] = Map(
+      "state.label" -> Some("SUBMITTED"),
+      "state.description" -> Some("Deposit is valid and ready for post-submission processing"),
+      "depositor.userId" -> Some(depositorUserID),
+      "datamanager.userId" -> Some(datamanager),
+      "datamanager.email" -> Some(emailaddress),
+      "springfield.domain" -> sf.map(_.domain),
+      "springfield.user" -> sf.map(_.user),
+      "springfield.collection" -> sf.map(_.collection)
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala:95]: 👍 without asking these 3 springfield properties nicely match my PR https://github.com/DANS-KNAW/easy-ingest-flow/pull/13/files#diff-a60d32db4073f1aac9e7ac6d6917c9d5

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala [hunk 3]
```diff
@@ -18,77 +18,101 @@ package nl.knaw.dans.easy.multideposit.actions
...
+import scala.util.{ Failure, Try }
 import scala.xml.Elem
 
 case class AddFileMetadataToDeposit(row: Int, entry: (DatasetID, Dataset))(implicit settings: Settings) extends UnitAction[Unit] {
 
   val (datasetID, dataset) = entry
 
+  // TODO @rvanheest move the mimetype computation to a separate action once the new
+  // parameterized version of Action is installed?
+  lazy val mimetypeMap: Try[List[(File, String)]] = Try {
```
Comment (PaulBoon) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala]: Maybe introduce type MimeType = String

File: multi-deposit-instructions.md [hunk 4]
```diff
@@ -108,24 +107,28 @@ The metadata elements starting with `SF_` are used to create a Streaming Surroga
...
-  URL of said Streaming Surrogate in Springfield. These fields may only be used if
+  URL of the Streaming Surrogate in Springfield. These fields may only be used if
   all of them are specified. Since the `SF_PRESENTATION` is not yet defined at this point in time,
-  a placeholder called '_presentation-placeholder_' is put in its place.
-* Files marked with `FILE_AUDIO_VIDEO` = "Yes" (see below) are added to this presentation
-  through the Springfield Inbox and an entry in the Springfield Actions file that 
-  will be created by `process-sip`.
-* `SF_SUBTITLES` specifies a file that contains the subtitles for the presentation 
-  specified.
+  a placeholder called '_$presentation-placeholder_' is put in its place.
```
Comment (jo-pol) [inline multi-deposit-instructions.md:115]: thought we should change that, another PR?

File: multi-deposit-instructions.md [hunk 5]
```diff
@@ -108,24 +107,28 @@ The metadata elements starting with `SF_` are used to create a Streaming Surroga
...
+* All files in a dataset that are identified as audio/video (using mimetype detection) are added
+  to this presentation by identifying them as such (and providing extra metadata) in `files.xml`.
+* The data provided in `SF_DOMAIN`, `SF_USER` and `SF_COLLECTION` are stored for further processing
+  in the `deposit.properties` file.
+* `SF_ACCESSIBILITY` provides a way to override the file-accessibility from the default
+   `DDM_ACCESSRIGHTS` for all A/V file.
+* The columns `AV_FILE`, `AV_FILE_TITLE`, `AV_SUBTITLES` and `AV_SUBTITLES_LANGUAGE` are used to
```
Comment (jo-pol) [inline multi-deposit-instructions.md:122]: will ingest-flow revcieve these values in files.xml?

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala [hunk 6]
```diff
@@ -85,11 +101,16 @@ object AddDatasetMetadataToDeposit {
...
+        // if AV_FILE_TITLE and/or AV_SUBTITLES are given, AV_FILE must be present as well
+        checkRequiredWithGroup(row, rowVals, List("AV_FILE"), List("AV_FILE_TITLE", "AV_SUBTITLES", "AV_SUBTITLES_LANGUAGE")),
+        checkRequiredWithGroup(row, rowVals, List("AV_SUBTITLES"), List("AV_SUBTITLES_LANGUAGE")),
 
         // check allowed value(s)
         // scheme
         checkValueIsOneOf(row, rowVals, "DCT_TEMPORAL_SCHEME", "abr:ABRperiode"),
         checkValueIsOneOf(row, rowVals, "DC_SUBJECT_SCHEME", "abr:ABRcomplex"),
+        checkValueIsOneOf(row, rowVals, "DDM_ACCESSRIGHTS", AccessCategory.values().map(_.toString): _*),
+        checkValueIsOneOf(row, rowVals, "SF_ACCESSIBILITY", FileAccessRights.values.map(_.toString).toList: _*),
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:115]: `: _*` huh?

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala [hunk 7]
```diff
@@ -18,77 +18,103 @@ package nl.knaw.dans.easy.multideposit.actions
...
+      val avFiles = mimetypes.filter {
+        case (_, mimetype) => (mimetype startsWith "video") || (mimetype startsWith "audio")
+      }
+
+      if (avFiles.nonEmpty)
+        checkColumnsAreNonEmpty(row, dataset, "SF_USER", "SF_COLLECTION")
+          .recoverWith {
+            case ActionException(r, msg, _) => Failure(ActionException(r,
+              s"$msg\ncause: these columns should contain values because audio/video files are " +
+                s"found:\n${ avFiles.map { case (file, _) => s" - $file" }.mkString("\n") }"))
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala:67]: Aren't the file right also important whether to create streaming or not? Can't I leave these columns empty if I wan't to archive them but have them accessible to none?

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala [hunk 8]
```diff
@@ -18,77 +18,103 @@ package nl.knaw.dans.easy.multideposit.actions
...
       }
   }
 
-  def datasetToFileXml(datasetID: DatasetID)(implicit settings: Settings): Try[Elem] = {
-    val inputDir = multiDepositDir(datasetID)
+  private def datasetToFileXml: Try[Elem] = {
+    mimetypeMap.map(fileXmls(_) match {
+      case Nil => <files xmlns:dcterms="http://purl.org/dc/terms/"/>
+      case files =>
+        // @formatter:off
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala:93]: Can't we configure the behaviour of the formatter? On the other hand, if some developer hasn't the desired option installed it still might change unintentionally sooner or later.

**Review by PaulBoon — APPROVED** [review]

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala [hunk 9]
```diff
@@ -35,66 +34,94 @@ case class AddPropertiesToDeposit(row: Int, entry: (DatasetID, Dataset))(implici
...
 
-  def writeProperties(row: Int, datasetID: DatasetID, dataset: Dataset, emailaddress: DatamanagerEmailaddress)(implicit settings: Settings): Try[Unit] = {
+  private def writeProperties(depositorUserID: String, emailaddress: DatamanagerEmailaddress)(implicit settings: Settings): Try[Unit] = {
     val props = new Properties {
       // Make sure we get sorted output, which is better readable than random
       override def keys(): ju.Enumeration[AnyRef] = Collections.enumeration(new ju.TreeSet[Object](super.keySet()))
     }
 
-    addProperties(props, dataset, settings.datamanager, emailaddress)
+    Try { addProperties(props, depositorUserID, settings.datamanager, emailaddress) }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala:78]: Does ingest-flow use encoding to rewrite ddm/properties?

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala [hunk 10]
```diff
@@ -35,66 +34,94 @@ case class AddPropertiesToDeposit(row: Int, entry: (DatasetID, Dataset))(implici
...
-  private def validateDepositor(dataset: Dataset) = {
+  private def getDepositorID(dataset: Dataset): Try[String] = {
     dataset.get("DEPOSITOR_ID")
       .flatMap(_.headOption)
       .map(Success(_))
       .getOrElse(Failure(new IllegalStateException("""The column "DEPOSITOR_ID" is not present""")))
   }
+
+  /**
+   * @return Retrieve the springfield data from the dataset iff the springfield data is present
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala]: typo: iff

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala [hunk 11]
```diff
@@ -35,66 +34,94 @@ case class AddPropertiesToDeposit(row: Int, entry: (DatasetID, Dataset))(implici
...
+   */
+  private def getSpringfieldData: Option[SpringfieldData] = {
+    for {
+      user <- dataset.findValue("SF_USER")
+      collection <- dataset.findValue("SF_COLLECTION")
+    } yield {
+      dataset.findValue("SF_DOMAIN")
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala:118]: what is so special about the domain that it is not retrieved before the yield? I vagualy remember special treatment elsewhere.

File: src/test/resources/allfields/output/input-ruimtereis03/bag/metadata/dataset.xml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    <dc:title>Title for empty dataset</dc:title>
+    <dcterms:description>
+      This is an entry for a dataset with no data in it
+    </dcterms:description>
+    <dcx-dai:creatorDetails>
+      <dcx-dai:author>
+        <dcx-dai:surname>Duck</dcx-dai:surname>
+        <dcx-dai:initials>Donald</dcx-dai:initials>
+        <dcx-dai:organization>
+          <dcx-dai:name xml:lang="en">Duckstad</dcx-dai:name>
```
Comment (jo-pol) [inline src/test/resources/allfields/output/input-ruimtereis03/bag/metadata/dataset.xml:14]: Duckburg

File: src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDepositSpec.scala [hunk 13]
```diff
@@ -46,24 +43,155 @@ class AddFileMetadataToDepositSpec extends UnitSpec with BeforeAndAfter with Bef
...
+      "SF_COLLECTION" -> List("", "")
+    )
+    inside(new AddFileMetadataToDeposit(1, (datasetID, dataset)).checkPreconditions) {
+      case Failure(ActionException(_, message, _)) =>
+        message should {
+          include("No values found for these columns: [SF_USER, SF_COLLECTION]") and
+            include("reisverslag/centaur.mpg") and
+            include("path/to/a/random/video/hubble.mpg") and
+            include("path/to/a/random/sound/chicken.mp3")
+        }
```
Comment (jo-pol) [inline src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDepositSpec.scala:89]: new syntax for me to apply in ingest-flow too.

File: src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDepositSpec.scala [hunk 14]
```diff
@@ -73,33 +201,29 @@ class AddFileMetadataToDepositSpec extends UnitSpec with BeforeAndAfter with Bef
...
+    AddFileMetadataToDeposit(1, (datasetID, dataset)).execute() shouldBe a[Success[_]]
 
-  it should "fail if the file does not exist" in {
-    inside(AddFileMetadataToDeposit.xmlPerPath(multiDepositDir(datasetID))(new File(multiDepositDir(datasetID), "reisverslag/deel01.doc"))) {
-      case Failure(e: FileNotFoundException) => e.getMessage should include ("reisverslag/deel01.doc")
+    Utility.trim(XML.loadFile(stagingFileMetadataFile(datasetID)))
+      .child
+      .map(node => (node \@ "filepath", node.child.filter(_.label == "format").head.text)) should {
+      have length 10 and
+        contain allOf(
```
Comment (jo-pol) [inline src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDepositSpec.scala:215]: another one to remember

File: src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDepositSpec.scala [hunk 15]
```diff
@@ -123,4 +123,47 @@ class AddPropertiesToDepositSpec extends UnitSpec with BeforeAndAfter with Befor
...
+    content should include ("state.label")
+    content should include ("state.description")
+
+    content should include ("depositor.userId=ruimtereiziger1")
+
+    content should include ("datamanager.email=dm@test.org")
+    content should include ("datamanager.userId=dm")
+    content should include ("springfield.domain=dans")
+    content should include ("springfield.user=janvanmansum")
+    content should include ("springfield.collection=Jans-test-files")
```
Comment (jo-pol) [inline src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDepositSpec.scala]: why here not the syntax I noticed above?

File: multi-deposit-instructions.md (hunk 4 above)
Comment (rvanheest) [reply multi-deposit-instructions.md:115]: Added this to EASY-1241

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala (hunk 6 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:115]: Indicate that the list has to be viewed as a varargs

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala (hunk 10 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala]: Nope, shorthand for if and only if (`<->`)

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala (hunk 11 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala:118]: The domain is optional and has a default value, which is stated in the class definition below.

File: src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDepositSpec.scala (hunk 15 above)
Comment (rvanheest) [reply src/test/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDepositSpec.scala]: Don't know. Can do that indeed...

File: multi-deposit-instructions.md (hunk 5 above)
Comment (rvanheest) [reply multi-deposit-instructions.md:122]: added extra explanation in 2b45707

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala (hunk 7 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddFileMetadataToDeposit.scala:67]: No, `SF_ACCESSIBILITY` is optional. It defaults to the file access right of `DDM_ACCESSRIGHTS`.

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala (hunk 10 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddPropertiesToDeposit.scala]: 😉  under hugh presure of @lindareijnhoudt I deleted the second `f` 😜

File: src/test/resources/allfields/output/input-ruimtereis03/bag/metadata/dataset.xml (hunk 12 above)
Comment (rvanheest) [reply src/test/resources/allfields/output/input-ruimtereis03/bag/metadata/dataset.xml:14]: last time I checked it was Duckstad 😄

**Pull request merged**